crates/ty_ide/src/semantic_tokens.rs

```diff
@@ -0,0 +1,2608 @@
+        let name_str = name.id.as_str();
+        let name_start = name.start();
+
+        // Split the dotted name and calculate positions for each part
+        let mut current_offset = ruff_text_size::TextSize::default();
+        for part in name_str.split('.') {
+            if !part.is_empty() {
+                self.add_token(
+                    ruff_text_size::TextRange::at(name_start + current_offset, part.text_len()),
+                    token_type,
+                    SemanticTokenModifier::empty(),
+                );
+            }
+            // Move past this part and the dot
+            current_offset += part.text_len() + '.'.text_len();
+        }
+    }
+
+    fn classify_from_alias_type(
+        &self,
+        ty: Type,
+        local_name: &ast::Identifier,
+    ) -> (SemanticTokenType, SemanticTokenModifier) {
+        self.classify_from_type_and_name_str(ty, local_name.id.as_str())
+    }
+
+    fn visit_type_annotation(&mut self, annotation: &ast::Expr) {
+        let prev_in_type_annotation = self.in_type_annotation;
+        self.in_type_annotation = true;
+        self.visit_expr(annotation);
+        self.in_type_annotation = prev_in_type_annotation;
+    }
+
+    // Visit parameters for a function or lambda expression and classify
+    // them as parameters, selfParameter, or clsParameter as appropriate.
+    fn visit_parameters(
+        &mut self,
+        parameters: &ast::Parameters,
+        func: Option<&ast::StmtFunctionDef>,
+    ) {
+        // Parameters
+        for (i, param) in parameters.args.iter().enumerate() {
+            let token_type = if let Some(func) = func {
+                // For function definitions, use the classification logic to determine
+                // whether this is a self/cls parameter or just a regular parameter
+                self.classify_parameter(&param.parameter, i == 0, func)
+            } else {
+                // For lambdas, all parameters are just parameters (no self/cls)
+                SemanticTokenType::Parameter
+            };
+
+            self.add_token(
+                param.parameter.name.range(),
+                token_type,
+                SemanticTokenModifier::empty(),
+            );
+
+            // Handle parameter type annotations
+            if let Some(annotation) = &param.parameter.annotation {
+                self.visit_type_annotation(annotation);
+            }
+        }
+    }
+
+    fn visit_body(&mut self, body: &[Stmt]) {
+        for stmt in body {
+            self.visit_stmt(stmt);
+        }
+    }
+}
+
+impl SourceOrderVisitor<'_> for SemanticTokenVisitor<'_> {
+    fn visit_stmt(&mut self, stmt: &Stmt) {
+        // If we have a range filter and this statement doesn't intersect, skip it
+        // as an optimization
+        if let Some(range_filter) = self.range_filter {
+            if stmt.range().intersect(range_filter).is_none() {
+                return;
+            }
+        }
+
```

Comment:
I suggest overriding `enter_node` instead and return `TraverseSignal::Skip` if a node shouldn't be visisted. This has the benefit that it also works for non-statements and avoids that nodes get (deeply) visited for which you don't override the `visit_` method.